pkg/retriever/cidretrieval.go

```diff
@@ -178,9 +177,24 @@ func (ret *retrieval) collectResults(candidates []RetrievalCandidate) (*Retrieva
 }
 
 // startRetrievals will begin async retrievals from the list of candidates
-func (ret *retrieval) startRetrievals(ctx context.Context, candidates []RetrievalCandidate) {
+func (ret *retrieval) startRetrievals(ctx context.Context, candidates []RetrievalCandidate, resultChan chan runResult, stopChan chan struct{}) {
 	for _, candidate := range candidates {
-		go ret.runRetrieval(ctx, candidate)
+		// go ret.runRetrieval(ctx, candidate, resultChan)
+		go func(candidate RetrievalCandidate) {
+			// try-receive stop channel to exit early
+			select {
```

Comment:
as I said below, I think these two selects are redundant and could be squashed into one; the first one isn't providing anything that the second one doesn't

I think the behaviour here would be that if would proceed through query->retrieval for all goroutines but only return the results for ones that were provided before `stopChan` was closed.

But tbh I don't really know what the behaviour is going to be if `stopChan` is closed before `runRetrieval` ends but I don't believe is going to be doing an early bail. I _think_ the second select is simply going to check that `stopChan` isn't closed and then proceed to run the blocking `runRetrieval` call until that returns. I might be wrong, but the lack of clarity on what's supposed to be going on is a warning flag I think. (Likewise, the fact that you found my original solution lacking is a warning flag too! So it probably should be addressed in some form, even if it's just clarification).